driver-core/src/main/com/mongodb/MongoClientSettings.java

```diff
@@ -0,0 +1,677 @@
+            writeConcern = settings.getWriteConcern();
+        }
+
+        /**
+         * Takes the settings from the given {@code ConnectionString} and applies them to the builder
+         *
+         * @param connectionString the connection string containing details of how to connect to MongoDB
+         * @return this
+         */
+        public Builder applyConnectionString(final ConnectionString connectionString) {
+            if (connectionString.getApplicationName() != null) {
+                applicationName = connectionString.getApplicationName();
+            }
+            ClusterSettings.Builder clusterSettingsBuilder = clusterSettings != null
+                    ? ClusterSettings.builder(clusterSettings) : ClusterSettings.builder();
+            clusterSettings = clusterSettingsBuilder.applyConnectionString(connectionString).build();
+            if (!connectionString.getCompressorList().isEmpty()) {
+                compressorList = connectionString.getCompressorList();
+            }
+            connectionPoolSettings = ConnectionPoolSettings.builder(connectionPoolSettings).applyConnectionString(connectionString).build();
+            if (connectionString.getCredential() != null) {
+                credential = connectionString.getCredential();
+            }
+            heartbeatSocketSettings = SocketSettings.builder(heartbeatSocketSettings).applyConnectionString(connectionString).build();
+            if (connectionString.getReadConcern() != null) {
+                readConcern = connectionString.getReadConcern();
+            }
+            if (connectionString.getReadPreference() != null) {
+                readPreference = connectionString.getReadPreference();
+            }
+            retryWrites = connectionString.getRetryWrites();
+            serverSettings = ServerSettings.builder(serverSettings).applyConnectionString(connectionString).build();
+            socketSettings = SocketSettings.builder(socketSettings).applyConnectionString(connectionString).build();
+            sslSettings = SslSettings.builder(sslSettings).applyConnectionString(connectionString).build();
+            if (connectionString.getWriteConcern() != null) {
+                writeConcern = connectionString.getWriteConcern();
+            }
+            return this;
+        }
+
+        /**
+         * Applies the {@link ClusterSettings.Builder} block and then sets the clusterSettings.
+         *
+         * @param block the block to apply to the ClusterSettings.
+         * @return this
+         * @see MongoClientSettings#getClusterSettings()
+         */
+        public Builder applyToClusterSettings(final Block<ClusterSettings.Builder> block) {
+            notNull("block", block);
+            ClusterSettings.Builder builder = clusterSettings == null ? ClusterSettings.builder()
+                    : ClusterSettings.builder(clusterSettings);
+            block.apply(builder);
+            clusterSettings = builder.build();
```

Comment:
This is going to cause problems, as it prevents applying incomplete cluster settings.  This throws an exception, for example:

        MongoClientSettings settings = MongoClientSettings.builder()
                .applyToClusterSettings(builder -> {
                    builder.description("Hello MongoDB!");
                })
                .applyConnectionString(new ConnectionString("mongodb://localhost"))
                .build();

```
Exception in thread "main" java.lang.IllegalArgumentException: hosts can not be null
	at com.mongodb.assertions.Assertions.notNull(Assertions.java:36)
	at com.mongodb.connection.ClusterSettings.<init>(ClusterSettings.java:446)
	at com.mongodb.connection.ClusterSettings.<init>(ClusterSettings.java:43)
	at com.mongodb.connection.ClusterSettings$Builder.build(ClusterSettings.java:260)
	at com.mongodb.MongoClientSettings$Builder.applyToClusterSettings(MongoClientSettings.java:183)
	at HelloMongoDBNew.main(HelloMongoDBNew.java:31)
```

I can think of two solutions:

1. Make sure there are sensible defaults for all settings, such that MongoClientSettings.builder.build() won't fail.  I think this is a good idea regardless.  The only one I can see now is ClusterSettings.hosts, but I'm not sure what to do about the default.  Currently `hosts` is null, which we can change to "localhost:27017", but `clusterConnectionMode` defaults to MULTIPLE, even if hosts size is 1.  This is not the same as MongoClientOptions.  So not sure what to do about this one.  Maybe we can default clusterConnectionMode to null in the builder, and set it to SINGLE if hosts.size == 1 (which would be the default) and MULTIPLE if hosts.size > 1.  It's only in the rare case where you'd have a single host and still want MULTIPLE, so we shouldn't optimize for that.
2. Store Builder instances in MongoClientSettings.Builder.  This seems more straightforward anyway, and you can always get from a settings to its equivalent builder when necessary

Reply:
If we just use builders tunderneath then this problem goes away.

Reply:
Done.

Reply:
Done.

ClusterSettings.hosts defaults to [new ServerAddress()].  Mode default now based on size of hosts. 

I thought about overriding it in async to stay Multiple - but when a user applies cluster settings via the `applySettings` method it would be overwritten anyway - we can't tell if it the mode was set by a user or not.